### DISCUSSION TITLE
Statically linking libsodium.a to a dynamic libraray mylib.so.

### DISCUSSION THREAD

**Discussion by derickrk38** [jedisct1/libsodium#1322] [General]
Hello all

So I am trying to create a standalone dynamic library mylib.so from mylib.cpp for distribution. I am trying to link libsodium.a statically to it. But I get an error. A little new to C++ compiling and building.

commands that i used to build mylib.so: 
g++ -std=c++11  -c -fPIC -o mylib.o mylib.cpp
g++ -std=c++11 -shared -o mylib.so  mylib.o /usr/local/lib/libsodium.a

Error:

/usr/bin/ld: /usr/local/lib/libsodium.a(libsodium_la-crypto_secretbox_easy.o): relocation R_X86_64_PC32 against symbol `crypto_secretbox_detached' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status

based on some suggestion I tried recompiling libsodium with fPIC
./configure CFLAGS=-fPIC
make

but I still get the same error

Please help me out

**Comment by jedisct1** [comment]
That looks more like a question on how to use the GNU linker than a question on libsodium :)

https://stackoverflow.com/questions/2649735/how-to-link-static-library-into-dynamic-library-in-gcc

`--whole-archive` might be what you need.

**Reply by derickrk38** [reply]
Thanks for the reply. I tried that. But --whole-archive doesn't exist for g++ maybe only for gcc.  But I found a solution 
I needed to recompile libsodium with these configuration ./configure --with-pic --disable-pie --enable-static. 
https://github.com/jedisct1/libsodium/issues/955#issue-612983704

Then 

g++ -std=c++11 -c -o mylib.o mylib.cpp
ar cr mylib.a mylib.o
g++ -std=c++11 -o my_program my_program .cpp mylib.a path/to/libsodium.a

Though it doesn't create a standalone mylib.a , my final goal was to create a standalone executable which it creates. 

But I will try once with gcc compiler , may be I can build a standalone .a file
